cached.go

```diff
@@ -1,21 +1,142 @@
 package sseserver
 
 import (
+	"context"
 	"errors"
 	"fmt"
 	"net/http"
 	"sync"
 	"time"
-
-	cache "github.com/patrickmn/go-cache"
 )
 
+// cache is a custom cache implementation for events caching.
+type cache struct {
+	mu        sync.RWMutex
+	items     map[string]map[string]*Event
+	expirator struct {
+		ttl  time.Duration
+		intv time.Duration
+		list []ttlRef
+	}
+}
+
+// ttlRef contains information about a specific event and its expiration
+// information.
+type ttlRef struct {
+	topic      string
+	id         string
+	expiration time.Time
+}
+
+// newCache creates a new instance of a cache.
+func newCache(ttl, intv time.Duration) *cache {
+	c := &cache{
+		items: make(map[string]map[string]*Event),
+	}
+
+	c.expirator.ttl = ttl
+	c.expirator.intv = intv
+
+	return c
+}
+
+// cleanUp starts a blocking process of a cache clean up. Periodically it
+// checks expirator's item's ttl references and deletes expired items from the
+// topic's. A context can be used to stop the process.
+func (c *cache) cleanUp(ctx context.Context) {
+	tm := time.NewTimer(c.expirator.intv)
+	defer tm.Stop()
+
+	var tstamp time.Time
+	for {
+		select {
+		case tstamp = <-tm.C:
+		case <-ctx.Done():
+			return
+		}
+
+		c.mu.Lock()
+		var total int
+		for _, item := range c.expirator.list {
+			if item.expiration.After(tstamp) {
+				break
+			}
+
+			delete(c.items[item.topic], item.id)
+			if len(c.items[item.topic]) == 0 {
+				delete(c.items, item.topic)
+			}
+
+			total++
+		}
+
+		c.expirator.list = c.expirator.list[total:]
+		c.mu.Unlock()
+
+		tm.Reset(c.expirator.intv)
+	}
+}
+
+// add adds the event to the topic, mapping it to the specified id.
+func (c *cache) add(topic, id string, event *Event) {
```

Comment:
It seems that if you happened to add an item with an existing `id` it's `expiration` would not be extended

Reply:
I think we shouldn't use the same id for the SSE events. It creates two separate links, which can skip the updates if the resync is beginning before the updated id. 

For example, before the update, the `b` would point to the `c`, `d`, `e`, `f` and then `g`, where after update, the the `b` will point to the `c`, which will point to nothing and will be the last server id, skipping `d`, `e`, `f` and `g` updates.
```
 ┌───────┐       ┌───────┐       ┌───────┐       ┌───────┐       ┌───────┐       ┌───────┐       ┌───────┐
 │a      │       │b      │       │c      │       │d      │       │e      │       │f      │       │g      │
 │       ├──────►│       ├──────►│       ├──────►│       ├──────►│       ├──────►│       ├──────►│       │
 │       │       │       │       │       │       │       │       │       │       │       │       │       │
 └───────┘       └───────┘       └───────┘       └───────┘       └───────┘       └───────┘       └───────┘

 ┌───────┐       ┌───────┐       ┌───────┐       ┌───────┐       ┌───────┐       ┌───────┐       ┌───────┐       ┌───────┐
 │a      │       │b      │       │c      │       │d      │       │e      │       │f      │       │g      │       │c      │
 │       ├──────►│       ├──────►│       │       │       ├──────►│       ├──────►│       ├──────►│       ├──────►│       │
 │       │       │       │       │       │       │       │       │       │       │       │       │       │       │       │
 └───────┘       └───────┘       └───────┘       └───────┘       └───────┘       └───────┘       └───────┘       └───────┘
```

This test doesn't fail (see the order which was published and which was received) on master and this branch:
![](https://i.imgur.com/9D3cZSy.png)

Reply:
We should keep this in mind and I think it would be best if we could at least log an error when someone misuses the cache in such a way.